File: jar/src/main/java/jline/lib/butools/reptrans/TransformToMonocyclic.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 * Ported from BUTools V2.0
 */
package jline.lib.butools.reptrans;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import org.apache.commons.math3.complex.Complex;

import jline.util.Pair;
import jline.util.matrix.Matrix;

public final class TransformToMonocyclic {
    private TransformToMonocyclic() {}

    /**
     * Data class representing a Feedback Erlang Block (FEB).
     */
    private static final class FEB {
        final double lambda;
        final double z;
        final int n;
        final int multip;
        final List<Complex> evals;
        final List<Integer> emuls;

        FEB(double lambda, double z, int n, int multip, List<Complex> evals, List<Integer> emuls) {
            this.lambda = lambda;
            this.z = z;
            this.n = n;
            this.multip = multip;
            this.evals = evals;
            this.emuls = emuls;
        }
    }

    /**
     * Transforms an arbitrary matrix to a Markovian monocyclic matrix.
     */
    public static Matrix transformToMonocyclic(Matrix A, int maxSize, double precision) {
        Pair<List<Complex>, List<Integer>> evalsResult = eigvalc(A, precision);
        List<Complex> evalues = evalsResult.getLeft();
        List<Integer> repeats = evalsResult.getRight();

        List<FEB> febs = generateFEBs(evalues, repeats, maxSize, precision);

        int totalSize = 0;
        for (FEB feb : febs) {
            totalSize += feb.n * feb.multip;
        }

        Matrix B = Matrix.zeros(totalSize, totalSize);
        int pos = 0;
        for (int i = 0; i < febs.size(); i++) {
            int Ni = febs.get(i).n * febs.get(i).multip;
            Matrix febGen = febGenerator(febs.get(i).lambda, febs.get(i).z, febs.get(i).n, febs.get(i).multip);

            for (int r = 0; r < Ni; r++) {
                for (int c = 0; c < Ni; c++) {
                    B.set(pos + r, pos + c, febGen.get(r, c));
                }
            }

            if (i < febs.size() - 1) {
                double rowSum = 0.0;
                for (int c = 0; c < totalSize; c++) {
                    rowSum += B.get(pos + Ni - 1, c);
                }
                B.set(pos + Ni - 1, pos + Ni, -rowSum);
            }
            pos += Ni;
        }

        return B;
    }

    public static Matrix transformToMonocyclic(Matrix A, int maxSize) {
        return transformToMonocyclic(A, maxSize, 1e-14);
    }

    public static Matrix transformToMonocyclic(Matrix A) {
        return transformToMonocyclic(A, 100, 1e-14);
    }

    /**
     * Computes eigenvalues and their algebraic multiplicity.
     */
    private static Pair<List<Complex>, List<Integer>> eigvalc(Matrix A, double precision) {
        double tol = Math.sqrt(Math.ulp(1.0));

        List<Complex> eigArr = A.eig();
        List<Complex> eigenvalues = new ArrayList<Complex>(eigArr);
        Collections.sort(eigenvalues, new Comparator<Complex>() {
            @Override
            public int compare(Complex a, Complex b) {
                return Double.compare(a.getReal(), b.getReal());
            }
        });

        for (int i = 0; i < eigenvalues.size(); i++) {
            Complex rounded = new Complex(
                    Math.round(eigenvalues.get(i).getReal() / tol) * tol,
                    Math.round(eigenvalues.get(i).getImaginary() / tol) * tol);
            eigenvalues.set(i, rounded);
        }

        List<Complex> uniqueEvals = new ArrayList<Complex>();
        List<Integer> repeats = new ArrayList<Integer>();

        for (Complex ev : eigenvalues) {
            boolean found = false;
            for (int j = 0; j < uniqueEvals.size(); j++) {
                if (ev.subtract(uniqueEvals.get(j)).abs() <= tol) {
                    repeats.set(j, repeats.get(j) + 1);
                    found = true;
                    break;
                }
            }
            if (!found) {
                uniqueEvals.add(ev);
                repeats.add(1);
            }
        }

        return new Pair<List<Complex>, List<Integer>>(uniqueEvals, repeats);
    }

    /**
     * Generates Feedback Erlang Blocks from eigenvalues.
     */
    private static List<FEB> generateFEBs(List<Complex> evalues, List<Integer> repeats, int maxSize, double precision) {
        List<FEB> febs = new ArrayList<FEB>();
        int i = 0;
        int size = 0;

        while (i < evalues.size()) {
            int multip = repeats.get(i);
            double evalimag = -Math.abs(evalues.get(i).getImaginary());

            if (-evalimag < precision) {
                int n = 1;
                double sigma = -evalues.get(i).getReal();
                double z = 0.0;
                List<Complex> ev = new ArrayList<Complex>();
                ev.add(new Complex(evalues.get(i).getReal(), 0.0));
                List<Integer> em = new ArrayList<Integer>();
                em.add(multip);
                size += 1;

                febs.add(new FEB(sigma, z, n, multip, ev, em));
                i++;
            } else {
                int n = 3;
                size += 3;
                // The feedback probability z is strictly less than 1 only if the
                // ratio stays strictly below cot(pi/n). At the exact boundary
                // (ratio == cot(pi/n)) rounding can let the loop exit with z == 1,
                // which yields a block with zero exit rate, hence a degenerate
                // generator with a zero eigenvalue. Keep a margin so that the
                // boundary is always resolved upwards, as in MATLAB
                // TransformToMonocyclic.m.
                double boundaryMargin = Math.sqrt(Math.ulp(1.0));
                while (evalimag / evalues.get(i).getReal() >= 1.0 / Math.tan(Math.PI / n) - boundaryMargin) {
                    n++;
                    size++;
                    if (size > maxSize) {
                        throw new IllegalArgumentException(
                                "The representation is too large (>maxSize). No result returned.");
                    }
                }

                double sigma = -(2 * evalues.get(i).getReal()
                        + evalimag * (1.0 / Math.tan(Math.PI / n) - Math.tan(Math.PI / n))) / 2;
                double z = Math.pow(
                        -evalimag * (1.0 / Math.tan(Math.PI / n) + Math.tan(Math.PI / n)) / (2 * sigma),
                        n);

                List<Complex> ev = new ArrayList<Complex>();
                List<Integer> em = new ArrayList<Integer>();
                for (int k = 0; k < n; k++) {
                    double zRoot = Math.pow(z, 1.0 / n);
                    double angle = 2 * k * Math.PI / n;
                    ev.add(new Complex(
                            -(1 - zRoot * Math.cos(angle)) * sigma,
                            zRoot * Math.sin(angle) * sigma));
                    em.add(multip);
                }

                febs.add(new FEB(sigma, z, n, multip, ev, em));
                i += 2;
            }
        }

        Collections.sort(febs, new Comparator<FEB>() {
            @Override
            public int compare(FEB a, FEB b) {
                Complex aMax = null;
                for (Complex c : a.evals) {
                    if (aMax == null || c.abs() > aMax.abs()) aMax = c;
                }
                Complex bMax = null;
                for (Complex c : b.evals) {
                    if (bMax == null || c.abs() > bMax.abs()) bMax = c;
                }
                double aReal = (aMax != null) ? aMax.getReal() : Double.MIN_VALUE;
                double bReal = (bMax != null) ? bMax.getReal() : Double.MIN_VALUE;
                return Double.compare(bReal, aReal);
            }
        });

        return febs;
    }

    /**
     * Generates the generator matrix for a single FEB.
     */
    private static Matrix febGenerator(double lambda, double z, int n, int multip) {
        int size = multip * n;
        Matrix A = Matrix.zeros(size, size);

        for (int ii = 0; ii < multip; ii++) {
            for (int j = 0; j < n; j++) {
                A.set(ii * n + j, ii * n + j, -lambda);
                if (j < n - 1) {
                    A.set(ii * n + j, ii * n + j + 1, lambda);
                }
            }
            A.set(ii * n + n - 1, ii * n,
                    A.get(ii * n + n - 1, ii * n) + z * lambda);

            if (ii < multip - 1) {
                A.set(ii * n + n - 1, (ii + 1) * n, (1 - z) * lambda);
            }
        }

        return A;
    }
}

File: jar/src/main/kotlin/jline/lib/butools/reptrans/TransformToMonocyclic.kt
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 * Ported from BUTools V2.0
 *
 * Reference:
 * Mocanu, S., Commault, C.: "Sparse representations of phase-type distributions,"
 * Stoch. Models 15, 759-778 (1999)
 */
package jline.lib.butools.reptrans

import jline.util.matrix.Matrix
import org.apache.commons.math3.complex.Complex
import kotlin.math.*

/**
 * Data class representing a Feedback Erlang Block (FEB).
 */
private data class FEB(
    val lambda: Double,
    val z: Double,
    val n: Int,
    val multip: Int,
    val evals: List<Complex>,
    val emuls: List<Int>
)

/**
 * Transforms an arbitrary matrix to a Markovian monocyclic matrix.
 *
 * @param A Matrix parameter of the initial representation (shape N,N)
 * @param maxSize The maximal order of the resulting Markovian representation (default 100)
 * @param precision Matrix entries smaller than the precision are considered to be zeros (default 1e-14)
 * @return Transient generator matrix of the Markovian monocyclic representation.
 *         Note that M>N if there are complex eigenvalues.
 * @throws IllegalArgumentException if no Markovian monocyclic generator has been found up to the given size.
 */
fun transformToMonocyclic(A: Matrix, maxSize: Int = 100, precision: Double = 1e-14): Matrix {
    // Determine eigenvalues and their multiplicities
    val (evalues, repeats) = eigvalc(A, precision)

    // Build monocyclic representation
    val febs = generateFEBs(evalues, repeats, maxSize, precision)

    // Calculate total size
    var totalSize = 0
    for (feb in febs) {
        totalSize += feb.n * feb.multip
    }

    // Assemble generator matrix of the hyper-feb
    val B = Matrix.zeros(totalSize, totalSize)
    var pos = 0
    for (i in febs.indices) {
        val Ni = febs[i].n * febs[i].multip
        val febGen = febGenerator(febs[i].lambda, febs[i].z, febs[i].n, febs[i].multip)

        for (r in 0 until Ni) {
            for (c in 0 until Ni) {
                B[pos + r, pos + c] = febGen[r, c]
            }
        }

        if (i < febs.size - 1) {
            var rowSum = 0.0
            for (c in 0 until totalSize) {
                rowSum += B[pos + Ni - 1, c]
            }
            B[pos + Ni - 1, pos + Ni] = -rowSum
        }
        pos += Ni
    }

    return B
}

/**
 * Computes eigenvalues and their algebraic multiplicity.
 */
private fun eigvalc(A: Matrix, precision: Double): Pair<List<Complex>, List<Int>> {
    val tol = sqrt(Math.ulp(1.0))  // sqrt(machine epsilon), matches MATLAB's sqrt(eps)

    val eigenvalues = A.eig().toMutableList()
    eigenvalues.sortBy { it.real }

    // Round eigenvalues
    for (i in eigenvalues.indices) {
        val rounded = Complex(
            (eigenvalues[i].real / tol).roundToLong() * tol,
            (eigenvalues[i].imaginary / tol).roundToLong() * tol
        )
        eigenvalues[i] = rounded
    }

    // Get unique eigenvalues and their multiplicities
    val uniqueEvals = mutableListOf<Complex>()
    val repeats = mutableListOf<Int>()

    for (ev in eigenvalues) {
        var found = false
        for (j in uniqueEvals.indices) {
            if (ev.subtract(uniqueEvals[j]).abs() <= tol) {
                repeats[j] = repeats[j] + 1
                found = true
                break
            }
        }
        if (!found) {
            uniqueEvals.add(ev)
            repeats.add(1)
        }
    }

    return Pair(uniqueEvals, repeats)
}

/**
 * Generates Feedback Erlang Blocks from eigenvalues.
 */
private fun generateFEBs(evalues: List<Complex>, repeats: List<Int>, maxSize: Int, precision: Double): List<FEB> {
    val febs = mutableListOf<FEB>()
    var i = 0
    var size = 0

    while (i < evalues.size) {
        val multip = repeats[i]
        val evalimag = -abs(evalues[i].imaginary)

        if (-evalimag < precision) {
            // Real eigenvalue
            val n = 1
            val sigma = -evalues[i].real
            val z = 0.0
            val ev = listOf(Complex(evalues[i].real, 0.0))
            val em = listOf(multip)
            size += 1

            febs.add(FEB(sigma, z, n, multip, ev, em))
            i++
        } else {
            // Complex eigenvalue
            var n = 3
            size += 3
            while (evalimag / evalues[i].real >= 1.0 / tan(PI / n)) {
                n++
                size++
                if (size > maxSize) {
                    throw IllegalArgumentException("The representation is too large (>maxSize). No result returned.")
                }
            }

            val sigma = -(2 * evalues[i].real + evalimag * (1.0 / tan(PI / n) - tan(PI / n))) / 2
            val z = (-evalimag * (1.0 / tan(PI / n) + tan(PI / n)) / (2 * sigma)).pow(n)

            val ev = mutableListOf<Complex>()
            val em = mutableListOf<Int>()
            for (k in 0 until n) {
                val zRoot = z.pow(1.0 / n)
                val angle = 2 * (k) * PI / n
                ev.add(Complex(
                    -(1 - zRoot * cos(angle)) * sigma,
                    zRoot * sin(angle) * sigma
                ))
                em.add(multip)
            }

            febs.add(FEB(sigma, z, n, multip, ev, em))
            i += 2 // Skip conjugate pair
        }
    }

    // Order according to the dominant eigenvalue (descending)
    // MATLAB max() on complex arrays selects by magnitude, then < compares real parts
    return febs.sortedByDescending { feb -> feb.evals.maxByOrNull { it.abs() }?.real ?: Double.MIN_VALUE }
}

/**
 * Generates the generator matrix for a single FEB.
 */
private fun febGenerator(lambda: Double, z: Double, n: Int, multip: Int): Matrix {
    val size = multip * n
    val A = Matrix.zeros(size, size)

    for (ii in 0 until multip) {
        // Fill the n x n block
        for (j in 0 until n) {
            A[ii * n + j, ii * n + j] = -lambda
            if (j < n - 1) {
                A[ii * n + j, ii * n + j + 1] = lambda
            }
        }
        // Feedback connection
        A[ii * n + n - 1, ii * n] = A[ii * n + n - 1, ii * n] + z * lambda

        // Connection to next block
        if (ii < multip - 1) {
            A[ii * n + n - 1, (ii + 1) * n] = (1 - z) * lambda
        }
    }

    return A
}
